src/main/puzzle/milestone-manager.gd

```diff
@@ -0,0 +1,88 @@
+extends Node
+"""
+Tracks the player's progress towards milestones.
+
+Milestones can involve reaching a certain score, clearing a certain number of lines or surviving a certain number of
+seconds.
+
+Reaching a milestone can increase the level, end the scenario, or trigger a success condition.
+"""
+
+func _ready() -> void:
+	PuzzleScore.connect("score_changed", self, "_on_PuzzleScore_score_changed")
+
+
+func _physics_process(_delta: float) -> void:
+	if PuzzleScore.game_active and milestone_met(Scenario.settings.finish_condition):
+		PuzzleScore.end_game()
+
+
+"""
+Returns 'true' if the player has met the specified milestone.
+"""
+func milestone_met(milestone: Milestone) -> bool:
+	var result := false
+	var progress := milestone_progress(milestone)
+	match milestone.type:
+		Milestone.NONE:
+			result = false
+		Milestone.TIME_UNDER:
+			result = progress <= milestone.value
+		_:
+			result = progress >= milestone.value
+	return result
+
+
+"""
+Returns the player's current progress toward the specified milestone.
+"""
```

Comment:
Change comment to be more descriptive. This returns a 'float', what is it? Something in the range [0.0-1.0]? A number of lines, number of seconds?

Reply:
Improved comment.